### ISSUE TITLE
Extracting the title information from produced div

### ISSUE THREAD

**Issue by talatkuyuk** [koka831/remark-custom-container#141]
In order to create my custom component like `<Callout type={type} title={title}>{children}</Callout>`, the options `className` and `containerTag` are too handy, but also I needed two things to do:
+ to extract the title information from the div structure that produced by this plugin; 
+ to remove the inner title div if exists.

And, I face difficulty about playing with the div structure provided by this plugin in especially typescript and react since the react children can be `string | number | true | ReactElement<any, string | JSXElementConstructor<any>> | ReactFragment | ReactPortal | ReactNode[]`, and it is cumbersome to play with the type of children.

In order to make our job easier; **my proposal is** to give an option about the title, either:
+ **to produce the title `in inner div`** **(as a default, as it is in the current)** or,
+ **to provide the title info in the `title` property `of the first outer div` _without having the inner title div_**. 

```ts
use(container, {
  className: string, // default to "remark-container",
  containerTag: string, // default to "div"
  isTitleProperty?: boolean // optional like first two options; default to false  <------ my proposal
})
```

If the `isTitleProperty` is set the `true` then the plugin will produce the HTML structure like below:
```ts
<div class="remark-container className" title="Custom Title">
  Container Body (children)
</div>
```

if there is no title provided in the markdown, then:
```ts
<div class="remark-container className">
  Container Body (children)
</div>
```

If you approve my proposal, we as developers can **create a Custom component** in easy way by extracting the title info.

**Comment by koka831** [comment]
Hi @talatkuyuk, thank you for reporting the issue!
I like your idea and I think it's worth too, so how about make it more usable generally:)
In https://github.com/koka831/remark-custom-container/issues/138, we discussed about a render/transform function as an option, and now I propose two new transform APIs, `container` and `title`.
with these options, developer may choose high/low level API for their usage.

### interface

- `container` is an option to construct custom container element.
  - if not given, default to `constructContainer` in https://github.com/koka831/remark-custom-container/blob/9ff8e12f3244e69217a8b3a9e84836c48817c3dd/src/index.ts#L41-L55 .
- `title` is an option to construct custom title element.
  - if not given, default to `constructTitle` in https://github.com/koka831/remark-custom-container/blob/9ff8e12f3244e69217a8b3a9e84836c48817c3dd/src/index.ts#L57-L66
  - this is useful in your usecase, you may set `container` function and `title` to `null` so that you can avoid additional/default title element.

```javascript
use(container, {
  className?: string, // high-level API
  containerTag?: string, // high-level API
  // the new APIs / provide low-level options so that developer may create custom component.
  container?: (children: Node<Data>[], className: string, title?: string) => Parent,
  title?: (title: string) => Paragraph,
})
```

### example

So, in your usecase, add `title` attribute to the container element, looks like that:
(I added `title: title` line to the default renderer)

```javascript
use(container, {
  // override `title` to `null` to avoid duplicate title in the container
  title: null,
  container: (children, className, title) => {
    return {
      type: "container",
      children,
      data: {
        hName: "div", // maybe some other html tag,
        hProperties: {
          // `hProperties` accepts custom attributes to the container.
          // so you may pass `title` attribute like that.
          title: title,
          className: [ // here you may set className of the container element.
            "remark-container",
            className.toLowerCase()
          ],
        },
      },
    };
  },
})
```

then it will convert:

```markdown
::: warn custom title

inner content

:::
```

into:

```html
<div class="remark-container warn" title="custom title">
  <p>inner content</p>
</div>
```

## drawbacks

I think it's bit more general, but make it difficult to use too.
This is just my opinion so I' willing to discuss more detail, and hope to find a good drop-off point between versatility and practicality.

**Comment by talatkuyuk** [comment]
Thank you very much for elaborating the issue.

The last edit in your comment (return  the `data` object only) is sufficient for flexibility, I suppose.

**Comment by talatkuyuk** [comment]
> It's (just) out of curiosity, @talatkuyuk would you like to implement it?

Of course, I will open a PR.

**Comment by talatkuyuk** [comment]
@koka831 Hi, I've started the implement it.

But, I realized that the options `className` and `containerTag` are doing their jobs is very well.  The new option `container` is not doing other than overwriting both options apart from adding additional properties. And, the `hName` and the `hProperties` keys in the return `data` object is a little bit opinionated, error proned, and may cause unintended errors if not provided properly.

Instead of providing the `container` option, we can provide a similar callback function but for adding additional properties in order to give desired flexibility. 
```js
additionalProperties?: (className: string, title?: string) => Record<string, unknown>;
```
For example, if provided like below
```js
additionalProperties: (className, title) => {
  title: title,
  ["data-type"]: className.toLowercase()
};
```
the outer container HTML element is going to have additional "title" and "data-type" properties.

For the `title` option, I would like to name it as `titleElement?: boolean;` (default `true`), since it is an option for adding or removing inner HTML **div element** contains the title.

In this way, the breaking change is not going to be required by the users, giving the same flexibility level for the users.

**Comment by koka831** [comment]
> may cause unintended errors if not provided properly.

Ah, it's a good point. I think your suggestion looks compact and have enough flexibility.
One thing, `titleElement?` might be better to be `Record<string, unknown> | null` to accept additional properties and can be set to `null`.

> In this way, the breaking change is not going to be required by the users, giving the same flexibility level for the users.

I agree. Thank you for the suggestion, LGTM.

**Comment by talatkuyuk** [comment]
I opened a PR for the issue: #146

**Comment by koka831** [comment]
released v1.3.0 https://www.npmjs.com/package/remark-custom-container/v/1.3.0
